### ISSUE TITLE
Improve Editable Children

### ISSUE THREAD

**Issue by reyma24** [godotengine/godot-proposals#3248]
### Describe the project you are working on

An immersive sim.

### Describe the problem or limitation you are having in your project

I believe that the editable children feature has the potential to be one of the most important part of Godot's scene editor. If improved, it would finally stop all those requests for component systems / multiple inheritance / modularity systems.
Godot already has the perfect solution for code modularity - adding nodes. This works perfectly and is increadibly easy to understand and implement:
![components](https://user-images.githubusercontent.com/90044741/132025635-e7eaac5e-cc6b-4966-8947-5a0240db43fc.png)

**The problem is with editing them after instancing the scene.**
I found 3, in fact:

1. When the scene contains many nodes, the tree becomes cluttered and shows nodes that are not needed (more in [27828](https://github.com/godotengine/godot/issues/27828)):
![not_needed](https://user-images.githubusercontent.com/90044741/132025792-ab4b2103-0fcc-4b63-af8f-0df4c028474d.png)

2. You have to **right-click > select editable children** every single time, on every single node  (more in [22263](https://github.com/godotengine/godot/issues/22263)).
![editable](https://user-images.githubusercontent.com/90044741/132026191-52cff893-df0a-48e5-b922-04396a695aa2.png)

3. The current color of the properties (red) indicates that those properties should not be edited. I have not found a reason for this anywhere, please tell me if there is (see [this](https://godotengine.org/qa/102154/colored-properties-editor-editing-children-instanced-scene)).
![red](https://user-images.githubusercontent.com/90044741/132025977-4765a336-40d6-4db9-a101-568be94f0552.png)

There might be more problems with editable children, these are the ones that I found and think are important things to change / implement.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Improve editing children's properties in instanced scenes.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Solutions:
1. **Select which nodes will be editable** - there should be an option for setting a node's visiblity when parent node is instanced:
![editable_in_parent](https://user-images.githubusercontent.com/90044741/132026537-85134251-435c-4fc2-b614-050f41790493.png)

2. **Show editable nodes by default** - there should be an option in the Editor Settings for showing editable children by default (automatically, when node enters tree):
![new_editor_setting](https://user-images.githubusercontent.com/90044741/132026562-4b468acf-4dae-4c8e-a8ac-504ceda19449.png)

3. Remove the red coloring from childrens' properties.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

The alternative is to create an addon that shows the children attributes in the parent's inspector, however that also comes with many problems (editor doesn't save changes to non-editable children and you can't attach metadata to instanced scenes).

### Is there a reason why this should be core and not an add-on in the asset library?

The feature is already implemented, the only thing it needs are improvements listed above.

_The end result:_
![final](https://user-images.githubusercontent.com/90044741/132027429-5e6b4f6a-3669-4de6-b60d-afb092cea93e.png)

**Comment by Calinou** [comment]
> Remove the red coloring from childrens' properties.

I wouldn't remove this coloring as it indicates that your changes may be lost in the future. We can change it to the warning color (yellow) instead to make it less scary.

**Comment by NathanLovato** [comment]
I agree with Calinou here, the issue with Editable Children isn't the presentation right now: as with inherited scenes, you have to be extremely careful not to lose your changes.

At GDQuest, we very rarely use Editable Children because of that: it's too easy to change the parent scene or refactor a class and introduce game-breaking bugs you won't detect until days or weeks later.

With that limitation fixed somehow, sure, it'd be a great feature. There was also a proposal to improve inherited scenes and limit or prevent data losses. To me, that'd be the starting point.

**Comment by Wavesonics** [comment]
Maybe the data loss is really what we should focus on finding a solution for. At the very least, knowing that the data loss has happened (_like a bunch of changes visible in your source control_) so that it doesn't silently break things.

Can someone explain the root cause of the data loss?

**Comment by Calinou** [comment]
> Can someone explain the root cause of the data loss?

One common cause for this is when you rename a property.

**Comment by raulsntos** [comment]
> > Can someone explain the root cause of the data loss?
> 
> One common cause for this is when you rename a property. There's already an open proposal about adding an annotation to handle property renames, but I can't find it right now.

#3152

**Comment by TackerTacker** [comment]
I would like a solution that avoids showing the children completely, but it would still be better than what we have right now. 

In your _the end result_ image you show the "Attacker" child as selected and the properties you want from that child, the "Script Variables". 
The Scene tab can become quite cluttered if there are a lot of nodes. Instanced scenes help fight this by creating bigger building blocks out of smaller pieces which get hidden to de-clutter the scene graph, so this would counteract this by partially bringing back hidden away nodes, even though you don't actually want the hidden nodes back, you want to be able to change the values found under "Script Variables" and nothing else. 
I think it would be much better to make only selected properties, or property groups available directly on the instanced scene itself, without cluttering up the scene dock with child nodes.

**Comment by justinbarrett** [comment]
I am not a contributor, but I have never had a case where I needed to edit the children.. I have always just reopened the scene, fixed the issue and moved on...This is another reason there are exported variables. That is where I do any changes needed....if someone implements this...it's, ofc, fine...but I feel like this is the wrong path and might teach bad habits. Feels very "corner case".

**Comment by mark-rafter** [comment]
I would prefer a variant of this where you can selectively override children (by right clicking the node and checking "Override Child").. Currently I'm doing this in code by exporting NodePath but it's a bit tedious and pollutes the code/scene tree, and is susceptible to bugs (e.g. no static type checking).

### Current behaviour
```csharp
// Character.cs
[Export] public NodePath CharacterAnimationTreePath { get; set; } = nameof(CharacterAnimationTree);

public CharacterAnimationTree CharacterAnimationTree { get; private set; }

public override void _Ready()
{
    CharacterAnimationTree = GetNode<CharacterAnimationTree>(CharacterAnimationTreePath);
}
```
![override-current](https://user-images.githubusercontent.com/38855689/132243522-56aa3eb7-dd62-4f3c-90b3-f5775365893a.png)

### Proposed behaviour
```csharp
// Character.cs
public CharacterAnimationTree CharacterAnimationTree { get; private set; }

public override void _Ready()
{
    CharacterAnimationTree = GetNode<CharacterAnimationTree>(nameof(CharacterAnimationTree));
}
```
![override-new](https://user-images.githubusercontent.com/38855689/132243545-ba074d34-4bd8-4c0a-b8b3-12ebe1f2c54f.png)

**Comment by reyma24** [comment]
> I have never had a case where I needed to edit the children.. I have always just reopened the scene, fixed the issue and moved on...This is another reason there are exported variables. That is where I do any changes needed....if someone implements this...it's, ofc, fine...but I feel like this is the wrong path and might teach bad habits. Feels very "corner case".

**Here's my thinking, visualized with an example:**
You can't add **the same** _Destructible_ script to, for example, both Static and Kinematic bodies - they extend different classes. You can only do something like this (each is a different script):
![extend_from_below](https://user-images.githubusercontent.com/90044741/132251137-5dcab160-7273-43a1-81d9-e3d89b545663.png)


This is will quickly become unmanageable.
However, this way you will get all the properties needed in the inspector:
![all_properties](https://user-images.githubusercontent.com/90044741/132250572-ef9844c3-b39e-4206-82b2-5fa2fd6abcec.png)





### The much better solution is to "extend" them by adding the same node:
![the_same_script](https://user-images.githubusercontent.com/90044741/132250200-b8823d9a-918e-43a7-bd6d-3f900bbf9633.png)


Great! But now you can't see and edit those properties after instancing this scene because the inspector shows only the properties from the classes above:
![no_properties](https://user-images.githubusercontent.com/90044741/132250603-6e3053fc-3222-4f52-97a6-f37e9e2c6587.png)


The solution: **Editable Children**, hence this proposal.

**Comment by AttackButton** [comment]
> I am not a contributor, but I have never had a case where I needed to edit the children.. I have always just reopened the scene, fixed the issue and moved on...This is another reason there are exported variables. That is where I do any changes needed....if someone implements this...it's, ofc, fine...but I feel like this is the wrong path and might teach bad habits. Feels very "corner case".

This is not a "corner case", look at the likes in the op. I'm using this a lot in my project, in the level design. I'm using nodes from the base scene as a container for some other scenes that will be manipulated via script. The number and type of objects in the containers changes from instance to instance, as the level design ask by the context.

This feature is very important. Maybe creating something like `set_editable_instance(self)` and `set_display_folded(true or false)`.

**Comment by Z3R0PTR** [comment]
Hi.

I think, this will be a nice feature. In my projects, I have thousands of Models (FBX-Files). All of them are represented as inherited scenes, where the top node is a simple spatial (3D) that contains one ore more (hierarchical) mesh instances.
So if I want to instance a model multiple times within a scene and want to change the material for some of them, i have to edit the children and set the new material for all the mesh instances. This is my biggest problem in level design.

**Comment by eerbin13** [comment]
I am currently using a component based code architecture model. Improved Editable Children would be magnificent.

I stumbled across this thread looking for a way to turn Editable Children on by default.

Perhaps a method exposed to GDscript where you could use it in an _on_ready() with the tool keyword at the top of the script?

I was looking for a function inside Node called set_editable_children(state: bool) before I did my Google deep dive.

**Comment by MatthijsMud** [comment]
This could also be useful for creating "slots" if the various ancestor nodes aren't shown in the hierarchy. 

Granted, in case of [Spatial](https://docs.godotengine.org/en/stable/classes/class_spatial.html) and [Node2D](https://docs.godotengine.org/en/stable/classes/class_node2d.html) this could be worked around by creating nodes in the root that are targeted by [RemoteTransform](https://docs.godotengine.org/en/stable/classes/class_remotetransform.html) and [RemoteTransform2D](https://docs.godotengine.org/en/stable/classes/class_remotetransform2d.html) respectively. 

For [Control](https://docs.godotengine.org/en/stable/classes/class_control.html) nodes there currently exists no such alternative (that I'm aware of). Yet one might be interested in adding items to a "deeply" nested [Container](https://docs.godotengine.org/en/stable/classes/class_container.html) representing the content (of a custom container built out of different nodes). An accordion could be an example of such a component. 

This would allow viewing the results directly in the editor rather than having to rely on runtime (or `Tool` scripts) that move nodes around/create scene instances.

| Current editable children situation | With improvements to editable children |
|-|-|
|![image](https://user-images.githubusercontent.com/11519728/180975813-ade89ed6-11bf-464f-b559-c1abf344e98f.png "Scene tree showing deep hierarchy of a character, where a \"Sword\" is the child of \"Right hand\"") | ![image](https://user-images.githubusercontent.com/11519728/180977472-1f4b655a-94aa-4745-a1c8-6a91b61c5396.png "Scene tree showing only \"Left hand\" and \"Right hand\" as children of instanced scene as slots, where a \"Sword\" is the child of \"Right hand\"") |

**Comment by BlockyDK** [comment]
I'm not entirely sure if this fits here, but I've just run into a use case that I think touches upon this discussion.

Considering the node based structure of Godot, it seems most intuitive to me to give quick and easy control over which properties from children nodes will be visible from the root node when instanced in another scene. And of course while not actually exposing the children to the scene tree in order to keep everything modular and compact. While this functionality can be achieved using @tool, there might be a better way.

Consider this prefab I put together, it is made up of several parts, a stand, a rotating brace, a tilting lamp and finally a spotlight.

![EditableChildren_01](https://user-images.githubusercontent.com/42628327/188357052-e027e224-8940-4658-84aa-630726a5f08c.PNG)

in the tool script I allow the user to rotate and tilt the head of the lamp with a simple slider, and as this is custom behavior, the tool script made sense here. And it is all in one tidy little node.

![EditableChildren_03](https://user-images.githubusercontent.com/42628327/188358879-403b2098-6893-46c2-8f05-083692f282f1.PNG)

![EditableChildren_02](https://user-images.githubusercontent.com/42628327/188358885-073fcdf5-1a44-4116-8bc6-4c46fbb62588.PNG)

However, I also wanted the ability to control the properties of the spotlight. This ended up resulting in a rather long tool script just exposing all of the properties from the spotlight. At first I only exposed a few (as seen above),  but as time went on I found continually needed access to more and more of the light's properties.

In this scenario, I found myself wishing for the ability to easily select properties from the child and have them auto expose in the root when instantiated in another scene. Or even better, a once click solution to have all of the selected child's properties exposed. Two approaches came to mind, one where the use would right click on the property in the child and the option would pop up to expose that property in the root when instantiated.

![EditableChildren_04](https://user-images.githubusercontent.com/42628327/188358087-4f916142-73d5-49c9-9672-f2647269f89d.PNG)

Another would be to add a checkbox in the properties for the child (so basically a property inherited from a basic node) that would just expose everything from that child in the root when instantiated.

Then ideally, in the case of this prefab for instance, when I select the instantiated prefab, over in the inspector, I should just see a dropdown for the spotlight child and all of it's properties (provided the box is checked under the spotlights properties from within the actual prefab's scene. Obviously this property would have to be excluded when exposing all the others.

That is just my two cents, I think it would work well with Godot's node based workflow.

**Comment by ExquisiterEmil** [comment]
As a level editor, this is one of my main pain points of the Godot Engine. I have so many cases where I need to edit children, when making a level.

1. There is no way to make children editable per default when instancing a scene (either global or per scene resource)
2. There is no hotkey to make the children of the current node editable
3. There is no way to hide any children, that you don't want to be exposed


@justinbarrett You don't have to be a contributor to come across many use cases for this. The moment you need two instances of a StaticBody with different Collision Shape sizes, you need to make children editable.

When you constantly have to instance new scenes, make their children editable, expand an abysmal hierarchy and find the one thing you are supposed to edit, this is detrimental for the workflow. More clicks, more searching, more typing on the level editor's end, which could be avoided, by the person who designed the feature, is always bad.

As an improvement, I think the easiest improvement could be adding the hotkey. Then, the solutions from @reyma24 's proposal would be a boon for level editing. Finally, @BlockyDK 's selectively exposed properties can be nice, but that would limit it to properties in the inspector, compared to using editor viewport gizmos for e.g. transformations.

**Comment by ExquisiterEmil** [comment]
@realkotob cool, from a level design standpoint, I can see this being very useful already! As a system designer, you often don't want the entire child node hierarchy visible to the level designer, so I am still of the opinion that something like solution 1 from the original proposal should be implemented as well.

**Comment by Huraqan** [comment]
Noone seems to talk about the loss of data... I lost half a level because of editable children. I saved a scene of my level to load it in a new project and most of the polygons got reset, emptied... The saved scene lost most of the data. How is this not the issue?

**Comment by realkotob** [comment]
@ExquisiterEmil I fully agree with you that adding a system to expose properties selectively would be useful, I can work on that as a second step after the first PR is accepted :)

**Comment by realkotob** [comment]
@Huraqan The data loss is an important point but it is a completely different issue.

There is a different proposal for how to fix the data loss issue that I will try to start working on soon. https://github.com/godotengine/godot-proposals/issues/3152

**Comment by Huraqan** [comment]
The link you shared shows a proposal about export variables and has nothing
to do with the issue of losing data when saving and loading a scene using
editable children. I don't understand why you shared that link.

**Comment by realkotob** [comment]
@Huraqan That issue is part of the losing data problem with editable children in instances, but yes you are right it is not related to saving and loading a scene using editable children.

Is there an existing issue for that? Otherwise we should create one

**Comment by Huraqan** [comment]
I just searched the repository and couldn't find an issue related to my
problem. I'll try to reproduce it and open up an issue.

**Comment by FeldrinH** [comment]
If @reyma24's 'Editable In Parent' checkbox is ever implemented then I would propose that the data loss risk could be reduced by showing a warning/confirmation dialog whenever the user attempts to change a child node that is marked as editable in a way that would result in data loss. So for example if you attempt to rename an editable child the editor could show a popup that says 'Renaming this node will lose any changes made to its properties in instances of this scene' with 'Ok' and 'Cancel' buttons.

**Comment by FranciscoDiMarzo** [comment]
This would be extremely useful for creating reusable components without cluttering the interface. Either exposing the children directly or selected children's properties on the inspector itself for the parent node/scene

**Comment by 3lis4** [comment]
> I found myself wishing for the ability to easily select properties from the child and have them auto expose in the root when instantiated in another scene.

Being able, **in a scene**, to choose on every node which property should be shown on the root when this scene is instantiated in another scene seems very useful.

We could also add an annotation to be able to specify in the root script which child properties we want to expose. This would be the **script equivalent** of the editor UI "expose in parent" option.

In a similar issue we imagined a syntactic sugar to help with exposing child vars : https://github.com/godotengine/godot-proposals/issues/7803#issuecomment-1732402845
> I think this proposal might be interesting as a **syntactic sugar** to limit the amount of "useless" setters when many exported variables of the child nodes need to be exported as well on the parent node.
> 
> It could be an annotation exposing a child node's exported variable. Something like `@export_child(path_to_child_node:var_name)` in parent node's script.

This `@export_child` was supposed to be just a syntactic sugar replacing an **export var and its setter** in the root script, which has two drawbacks : 
* You have to make your root script a tool script if you want to see in-editor updates of the child export var
* You loose the special edition tools of the child export var (like for creating a polygon for a CollisionPoligon2D)

But `@export_child` could be great as the script equivalent of the editor "expose in parent" option.

There is also this interesting discussion on reddit: https://www.reddit.com/r/godot/comments/173zn9j/how_do_you_handle_encapsulation_and_passing_data/

**Comment by foolmoron** [comment]
Here's another use case in support of reyma's proposal 1 to mark nodes as editable in parent. I have a CollisionPolygon2D as part of a scene and I want to edit the polygon in each instance. If I just exposed the `polygon` property, I would have to edit the array manually, but if I can expose the entire node, then I get the nice editor controls.

Of course, the option to expose specific properties would be very useful too. But I think exposing nodes is more important since it covers all use cases.

![image](https://github.com/godotengine/godot-proposals/assets/1753413/7bf41598-8300-40e1-9217-a85194b3610d)

**Comment by cursegate** [comment]
I agree that exposing child nodes to edit is the true solution to this problem.

We can use the editor to compose scenes without writing code, which is awesome. The problem is that we can't give a scene an intuitive API without code. A designer can't even tell if a scene's children require dependencies without enabling Editable Children and scrolling through.

I don't think the solution should be script-level, because a child node *without* a script attached can still have dependencies. Also, you might want to enable this feature for one node, but disable it for another node with the same script.

I don't think "Editable in Parent" is technically accurate, though. Something like "Editable in Instance" or "Exposed Outside Scene" sounds better to me.

**Comment by tiger94un** [comment]
can any one make  trash node to take back for history changes

**Comment by SirLich** [comment]
Hey folks! In case anyone has interest, I've written a plugin that does @3lis4 describes. Here is the source: https://github.com/SirLich/godot-hoist

You can opt-in individual properties from children/grandchildren, and they will appear on the root node for editing. Due to how scene serialization works, it's still required for 'editable children' to be enabled, but you don't need to go into the children anymore.

**Comment by hsandt** [comment]
Looks good in combination with https://github.com/godotengine/godot-proposals/issues/7649 (this addresses applying scene instance override, and already has an open PR)
